### ISSUE TITLE
Rancher request - remove unhealthy hosts automatically

### ISSUE THREAD

**Issue by nhumrich** [rancher/rancher#3354]
I have noticed that whenever you have an "unhealthy host" for your environment, rancher refuses to start any new containers (even on new healthy hosts) until that host becomes healthy and or manually removed. 
This causes an issue with autoscaling. 
I can set  a container to "global" scaling, and autoscaling up will work great, but I can not autoscale down because then the host that gets autoscaled (through AWS) away, becomes "unhealthy" in rancher, and all the sudden the environment wont start any more containers unless I go manually remove the host. A common scenario is when trying to replace hosts with new hosts. 

It would be great if rancher could automatically remove/delete unhealthy hosts (maybe by some threshold I define)

**Comment by deniseschannon** [comment]
What do you mean by "refuses to start any new containers"? If your host becomes unhealthy AND it has a heath check on it, Rancher should be re-scheduling the containers on the unhealthy host onto a healthy host. Without the health check, Rancher will not schedule any new containers of existing services.

With a global scale. if you added another host, it should automatically start containers of the service onto any newly added hosts. Having an unhealthy host shouldn't affect that deployment. 

By design, Rancher is not planning to remove unhealthy hosts automatically, there are too many variables that could have caused the host to become unhealthy and we want the users to look into those issues to either reconcile what's wrong to get hte host healthy or determine to remove the host themselves.

**Comment by nhumrich** [comment]
How do add a health-check to a host? I wasn't aware you could do that. 

Also you say "With a global scale. if you added another host, it should automatically start containers of the service onto any newly added hosts. Having an unhealthy host shouldn't affect that deployment. " but that isnt what happens. If I terminate an ec2 host, it becomes unhealthy, autoscaling brings up another, and the new one has 0 containers on it. As soon as I delete the unhealthy one, the "global scale" containers start to go up on the new one, but not until I delete it. 

I understand not remove unhealthy hosts by default. I also like the idea of moving containers onto a healthy host if the host isnt responding, but currently that doesnt happen. If it did, that would be expectable, then I could just go delete the old ones every once in a while.

**Comment by deniseschannon** [comment]
Here's the health checks information in the UI.
http://docs.rancher.com/rancher/rancher-ui/applications/stacks/adding-services/#health-checks

Concepts of Rancher:
http://docs.rancher.com/rancher/concepts/health-checks/

Can you share your docker-compose.yml for your global service? Do you have any host labels to select which hosts or is any host in your environment? 

I just ran this test. Had 5 hosts in my environment and used a global scale with no scheduling rules. Terminated 1 of my hosts in AWS so that it caused the host to go into reconnecting state. Added a different host into my Rancher setup. As soon as it joined Rancher, the container from the global service started on the new host.

**Comment by nhumrich** [comment]
I thought you meant there was a way to do a health check on the host. I am already doing service health checks. When i did the test, I had just one host. I terminated it, and let autoscaling create another. Here is the docker compose:

```
docker-gc:
  labels:
    io.rancher.scheduler.global: 'true'
    io.rancher.container.start_once: 'true'
    service: docker-gc
  tty: true
  image: spotify/docker-gc
  volumes:
  - /var/run/docker.sock:/var/run/docker.sock
  - /etc:/etc
  stdin_open: true
logspit:
  environment:
    SYSLOG_HOST: sumo
  labels:
    io.rancher.scheduler.global: 'true'
  tty: true
  image: canopytax/logspit
  links:
  - sumo-collector:sumo
  volumes:
  - /var/run/docker.sock:/var/run/docker.sock
  stdin_open: true
loadbalancer:
  ports:
  - 30001:8080
  labels:
    io.rancher.scheduler.global: 'true'
    service: loadbalancer
  tty: true
  image: rancher/load-balancer-service
  stdin_open: true
sumo-collector:
  ports:
  - 514:514/tcp
  - 514:514/udp
  labels:
    io.rancher.scheduler.global: 'true'
    service: sumo-collector
  tty: true
  image: sumologic/collector:latest-syslog
  stdin_open: true
cronboss:
  environment:
    INTERVAL: '2'
    SELECTOR_LABEL: service=myservice
    UNIT: minutes
  labels:
    service: cronboss
  command:
  - something
  - useful
  - here
  tty: true
  image: canopytax/cronboss
  volumes:
  - /var/run/docker.sock:/var/run/docker.sock
  stdin_open: true
```

And I also created a demo. This demo starts after the first host is terminated and the second host has registered itself with rancher, you will see that the global services dont move over until the first host is deleted. https://youtu.be/yQaZzSD5JHY

**Comment by deniseschannon** [comment]
@nhumrich The issue that you're facing is most likely due to the fact that there is no network agent on your new host. 

It's a combination of this issue: https://github.com/rancher/rancher/issues/1877 and https://github.com/rancher/rancher/issues/2287#issuecomment-171799879

Basically, it can't detect that the health check is unhealthy to move the LB over to the new hosts cause there are no other network agents to confirm that the health check is failing.

**Comment by kaos** [comment]
I agree with the `status/wontfix`, but perhaps there could be a warning issued somewhere when there are too few hosts to have a reliable setup with regard to health checks.

Is it sufficient running two nodes in a environment, or should we aim for at least 3 nodes?

**Comment by demarant** [comment]
@kaos
As I understood thete must be at least 4 hosts in an envfironment to have the health checks work properly and  Service HA to work.
If one service one one host has issues, there must be 3 haproxy on other **hosts** that also confirm unhealthy.
@deniseschannon
Please confirm. If this is the case, I think it would be wise to update documentation and make this more clear. Since lot of users evaluating rancher just add one environment and 2 or 3 hosts to test.

**Comment by nhumrich** [comment]
This sounds kind of weird. It makes sense for a scaled service, but why would a global service depend on health checks in any way? If the service is set to global with no scheduling rules, shouldn't it pop up on*every* new host no matter what?

**Comment by tombusby** [comment]
Main problem I'm having is that the documentation gives a great description of health checks, but basically zero walk-through instructions on how to set them up. I can see how to add a healthcheck when I spin up a new container, but not how to add one to existing stacks/services/containers.

**Comment by deniseschannon** [comment]
You can not add health checks to existing services, but you can upgrade the service to add a health check.

Would you want walk through instructions on how to do it in the UI or through rancher-compose?

**Comment by tombusby** [comment]
Either of those would be ideal tbh. My main issue at the moment is that the
docs don't appear to explain how do set healthchecks up, only that they
exist.

**Comment by deniseschannon** [comment]
@tombusby I've expanded the docs on how to set them up. Please let me know if you have additional questions. http://docs.rancher.com/rancher/latest/en/cattle/health-checks/

**Comment by tombusby** [comment]
That's great, thanks a lot for that. I'll check it out shortly.

**Comment by tombusby** [comment]
Looks great, thanks a lot @deniseschannon

**Comment by ampedandwired** [comment]
For anyone interested I have just published a small Docker service that automatically removes hosts from Rancher that have been terminated in AWS: https://github.com/ampedandwired/rancher-reaper

**Comment by wmbutler** [comment]
I wrote a bash script to do the same. You must install the rancher cli and configure your rancher credentials. Honestly, I think it's crazy that rancher isn't smart enough to remove hosts that don't respond. I can't fathom the use case where people would want to see hosts that aren't communicating. At the bare minimum, rancher should have some preferences like:

- [ ] Deactivate disconnected hosts immediately
- [ ] Remove deactivated hosts immediately

If a host miraculously comes back to life, rancher should accept it back into the fold, but in most cases, my hosts disappear because my aws autoscale group dropped them as part of a ruleset.

```bash
#!/bin/bash
rancher hosts ls | grep disconnected | awk '{print $1}' > hosts.txt
mapfile -t hosts < hosts.txt

for h in "${hosts[@]}"
do
   rancher deactivate $h; rancher rm $h
   echo "rancher deactivate $h; rancher rm $h"
done
rm hosts.txt
```

**Comment by kaos** [comment]
@wmbutler 
>  I can't fathom the use case where people would want to see hosts that aren't communicating.

I think I can help you there. We have physical hosts in our environment. If they stop communicating, that's a serious issue that needs attention. If rancher simply removed those hosts, it would be much harder to monitor them.

**Comment by vincent99** [comment]
FWIW 1.5+ have a global setting and a per-environment override (in the API, not in the UI) to automatically remove hosts disconnected longer than a number of minutes.

**Comment by drpebcak** [comment]
@vincent99 can you point me to some documentation or something on how to configure this?
